enhancements/etcd/automated-backups-no-config.md

```diff
@@ -0,0 +1,325 @@
+
+### Workload
+
+In order to take backups for Etcd data store and cluster resources, a workload must run on the cluster while having access to Etcd and cluster resources.
+Since Etcd data store of an Openshift cluster runs on `Pod` similar to other workload, several constructs such as `Deployment`, `DaemonSet`, `StatefulSet` and `SideCar` could be used.
+
+This section discusses two approaches (i.e. `SideCar` and `DaemonSet`), which are implemented and tested, explaining the Pros and Cons of each method.  
+Before deciding upon these two methods, several other approaches have been investigated. See [Alternatives](#Alternatives).
+
+#### SideCar container within all Etcd-static-pod
+
+A sidecar container is being deployed within each etcd pod in Openshift cluster in order to take backups.
+
+Initially, the sidecar container within etcd pods is not running (i.e. disabled). However, creating a `Backup` CR with `name=default` enables the container.
+
+Once enabled, the backups are being taken on each master node according to the provided `schedule` within the CR. Moreover, the backups are being maintained according to the supplied retention policy.
+
+The backups are being stored within a `hostPath` volume. The volume is a directory within each master node's file system.
+
+The approach has been implemented and tested. Below are the Pros and Cons found.
+
+_**Pros**_
+
+As the container runs on the same pod as etcd, it requests backup through `localhost`, since the containers runs on the same pod. Backups are being stored on the `hostPath` volume, with is a Directory on the node's file system. This guarantees no network delays.
+
+_**Cons**_
+
+Etcd pod is a static-pod, and the availability of the etcd cluster is vital for cluster stability. As the Etcd cluster is being managed by the `Cluster-etcd-operator`, and a `Deployment` resource manages the _mirrored_ etcd pods, any changes in the pod manifest results into rolling out a new revision, which leads to a duration of `etcd-cluster` unavailability, till the new revision is rolled out for all etcd pods. Consequently, the `API-Server` can not process any write requests, and read requests are being handled; if any, from the API-Server cache.   
+
+The backup sidecar container is being `disabled` by default. Upon creating a `Backup` CR on an Openshift cluster, the `PeriodicBackupController` reacts by enabling the container. This leads to a change in etcd static pod manifest, and a new revision is being rolled out. As explained above, such action causes etcd-cluster and API-Server unavailability.
+
+Additionally, any error within the backup sidecar, the container goes into `CrashLoopBackOff`, which leads to the whole `Pod` reports `NotReady` status. Consequently, the whole `Etcd-cluster` can not server requests, nor the `API-Server`.
+
+
+#### DaemonSet manages an Etcd-backup pod within every control-plane node.
```

Comment:
We had some previous discussion about using the parallel execution semantics of cronjobs, how did you get on with that research activity?

Reply:
yes, I have tested it here https://github.com/openshift/cluster-etcd-operator/pull/1363

Reply:
Yes, I have already tested this here https://github.com/openshift/cluster-etcd-operator/pull/1363 

you also have commented on the PR :D :D

Reply:
Ok, do we need to update the details here with the results of that investigation then?

Reply:
Right, just checking that, I don't really see why it was dismissed?

My understanding of the situation we had was:
* You tried to use the cronjob with parallelism
* You got stuck because you needed a PVC
* I suggested you just change it to use hostpath for now
* ...

Somewhere on this EP we discussed adding a hostpath option as an alternative configuration for PVC, and I believe I was told that this was on the roadmap for the backup feature anyway?

If you were to use the cronjob approach with hostpath, and the parallelism and completions config, would that work to schedule the backups, and have them on the hostpath, without the need for creating this whole new DS approach?

Reply:
Well, the ultimate  goal of this work is to provide AutomatedBackup **without** user supplied configuration.

I recall during testing https://github.com/openshift/cluster-etcd-operator/pull/1363 on a live cluster, I got stuck with configuring PVC, running parallel jobs requires parallel PVC 

I also tried patching the scheduled pods with hostPath but I don't remember why it did not work out 

wdyt @deads2k @tjungblu @vrutkovs  ?

Reply:
if we use a hostPath (and we finally all agree) - then it might make sense to have a dedicated field for it

Reply:
I have suggested with @Elbehery privately, that we move from having 
```
spec:
  ... // Other fields
  pvcName: ...
```
To
```
spec:
  ... // Other fields
  storage:
    type: PVC | HostPath
    pvc: // Only valid when type is PVC, required when type is PVC
      name: ...
```
For the automated backup case, we can create the default Backup with the storage type as HostPath, and the operator can modify the way it makes the podSpec for the cronjob based on the storage type. That would allow us to avoid the "we need multiple PVCs" option, but also allow, should the end user desire, to reconfigure the default, automated backup away from using hostpath and to an off cluster PVC storage, at which point we go back to 1 execution of the job at a time, as with any other backup

Reply:
Why is `type` then needed anyway? Can't you just discriminate based on whether the key `pvc` or `hostPath` exists? This is less error-prone (no mismatch possible); it only needs to be checked that not both are provided at the same time. (sorry, just stumbled over this, please ignore my comment if it doesn't make sense semantically; my comment is only based on API syntax)

Reply:
We use [Discriminated Unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions) so that we have a clear intention from the end user, which of the options they intended. 

Non-discriminated unions in upstream kube have caused confusion over time, and now the convention is to lean towards discriminated unions. More verbose, but more intentional.

> This is less error-prone (no mismatch possible)

We use CEL to ensure that the type, and non-empty union member are paired, so, there's no way for an end user to persist `pvc: {}` with `type: HostPath`.

Reply:
Thanks. I didn't know that there was already a convention about this. Personally, I think stating the top-level key is intention "enough" for non-squashed configuration groups, but I guess there has already been tons of discussion around this, so I for sure don't question the consensus here.

Sorry for my ignorance and the noise ...